monitor/dns_monitor.go

```diff
@@ -0,0 +1,177 @@
+package monitor
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/9corp/9volt/util"
+	log "github.com/Sirupsen/logrus"
+	resolver "github.com/miekg/dns"
+)
+
+const (
+	DEFAULT_DNS_TIMEOUT     = time.Duration(2) * time.Second
+	DEFAULT_DNS_RECORD_TYPE = "A"
+)
+
+type DnsMonitor struct {
+	Base
+	Timeout    time.Duration
+	Client     *resolver.Client
+	Expect     *regexp.Regexp
+	RecordType string
+}
+
+func NewDnsMonitor(rmc *RootMonitorConfig) *DnsMonitor {
+	dns := &DnsMonitor{
+		Base: Base{
+			RMC:        rmc,
+			Identifier: "dns",
+		},
+		Client:     &resolver.Client{},
+		Timeout:    DEFAULT_DNS_TIMEOUT,
+		RecordType: "A",
+	}
+
+	// Override the default with the setting if we have one
+	if rmc.Config.Timeout != util.CustomDuration(0) {
+		dns.Timeout = time.Duration(rmc.Config.Timeout)
+	}
+
+	// We'll just use the cumulative timeout
+	// https://godoc.org/github.com/miekg/dns#Client
+	dns.Client.Timeout = dns.Timeout
+
+	if len(rmc.Config.DnsRecordType) > 0 {
+		dns.RecordType = rmc.Config.DnsRecordType
+	}
+
+	if len(rmc.Config.Expect) > 0 {
+		dns.Expect = regexp.MustCompile(rmc.Config.Expect)
+	} else {
+		dns.Expect = regexp.MustCompile(".") // Match any non-empty
+	}
+
+	dns.MonitorFunc = dns.dnsCheck
+
+	return dns
+}
+
+func (dns *DnsMonitor) Validate() error {
+	log.Debugf(
+		"%v: Performing monitor config validation for %v",
+		dns.Identifier, dns.RMC.ConfigName,
+	)
+
+	if len(dns.RMC.Config.DnsTarget) < 1 {
+		return fmt.Errorf("No DNS target configured!")
+	}
+
+	if dns.Timeout >= time.Duration(dns.RMC.Config.Interval) {
+		return fmt.Errorf(
+			"'timeout' (%v) cannot equal or exceed 'interval' (%v)",
+			dns.Timeout.String(), dns.RMC.Config.Interval.String(),
+		)
+	}
+
+	if len(dns.RMC.Config.Expect) > 0 {
+		_, err := regexp.Compile(dns.RMC.Config.Expect)
+		if err != nil {
+			return fmt.Errorf("Unable to compile DNS check regexp! (%s)", err.Error())
+		}
+	}
+
+	return nil
+}
+
+// Do a full DNS check on a particular hostname against a particular
+// DNS server. In this check 'target' is the thing we're looking up
+// and 'host' is the DNS server we're talking to do the looking.
+func (dns *DnsMonitor) dnsCheck() error {
+	msg := &resolver.Msg{}
+
+	qType, ok := resolver.StringToType[strings.ToUpper(dns.RecordType)]
```

Comment:
Yeah, I couldn't believe it didn't, but the docs are not great and it took hours to find it.